tracker-neuralnet/ftnoir_tracker_neuralnet.cpp

```diff
@@ -356,80 +587,138 @@ std::optional<PoseEstimator::Face> PoseEstimator::run(
+    // latency spike of maybe an additional 50%. But it only occurs when the user
+    // moves his head far enough - or when the tracking ist lost ...
+    if (!last_localizer_roi_ || !last_roi_ ||
+        iou(*last_localizer_roi_,*last_roi_)<0.25)
     {
-        auto [p, rect] = localizer->run(grayscale);
-        last_inference_time += localizer->last_inference_time_millis();
-        if (p > 0.5 || rect.height < 5 || rect.width < 5)
+        auto [p, rect] = localizer_->run(grayscale_);
+        inference_time += localizer_->last_inference_time_millis();
+        
+        if (last_roi_ && iou(rect,*last_roi_)>=0.25 && p > 0.5)
+        {
+            // The new ROI matches the result from tracking, so the user is
+            // still there and to not disturb recurrent models, we only update
+            // ...
+            last_localizer_roi_ = rect;
+        }
+        else if (p > 0.5 && rect.height > 32 && rect.width > 32)
         {
-            last_localizer_roi = rect;
-            last_roi = rect;
+            // Tracking probably got lost since the ROI's don't match, but the
+            // localizer still finds a face, so we use the ROI from the localizer
+            last_localizer_roi_ = rect;
+            last_roi_ = rect;
         }
         else
         {
-            last_roi.reset();
-            last_localizer_roi.reset();
+            // Tracking lost and no localization result. The user probably can't be seen.
+            last_roi_.reset();
+            last_localizer_roi_.reset();
         }
     }
 
-    if (!last_roi)
+    if (!last_roi_)
     {
-        draw_gizmos(frame, {}, {});
+        draw_gizmos({}, {});
         return false;
     }
 
-    auto face = poseestimator->run(grayscale, *last_roi);
-    last_inference_time += poseestimator->last_inference_time_millis();
-    
+    auto face = poseestimator_->run(grayscale_, *last_roi_);
+    inference_time += poseestimator_->last_inference_time_millis();
+
     if (!face)
     {
-        last_roi.reset();
-        draw_gizmos(frame, *face, {});
+        last_roi_.reset();
+        draw_gizmos(*face, {});
         return false;
     }
 
-    last_roi = face->box;
+    {
+        // Here: compute ROI from head size estimate. This helps make the size prediction more
+        //       invariant to mouth opening. The tracking can be lost more often at extreme
+        //       rotations, depending on the implementation details. The code down here has
+        //       been tweaked so that it works pretty well.
+        // In old behaviour  ROI is taken from the model outputs
+        const vec3 offset = rotate_vec(face->rotation, vec3{0.f, 0.1f*face->size, face->size*0.3f});
+        const float halfsize = face->size/float(settings_.roi_zoom);
+        face->box = cv::Rect2f(
+            face->center.x + offset[0] - halfsize,
+            face->center.y + offset[1] - halfsize,
+            halfsize*2.f,
+            halfsize*2.f
+        );
+    }
```

Comment:
@DaWelter This is a change I'm not so sure about ... I might decide to revert it or make it an option. Depends on what works best for future networks.

Reply:
@sthalik , from my side you can merge it.